node/src/aura_or_nimbus_consensus.rs

```diff
@@ -0,0 +1,194 @@
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Manta is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Manta.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Implements a consensus that can  verify with both Nimbus and Aura
+//! NOTE: Nimbus is used for proposing exclusively
+
+use futures::TryFutureExt;
+use log::debug;
+use sc_client_api::HeaderBackend;
+use sc_consensus::{
+    import_queue::{BasicQueue, Verifier as VerifierT},
+    BlockImport, BlockImportParams,
+};
+use sc_consensus_aura::BuildVerifierParams;
+use sc_consensus_slots::InherentDataProviderExt;
+use sc_telemetry::TelemetryHandle;
+use session_key_primitives::aura::AuraId;
+use sp_api::ProvideRuntimeApi;
+use sp_block_builder::BlockBuilder as BlockBuilderApi;
+use sp_blockchain::Result as ClientResult;
+use sp_consensus::{error::Error as ConsensusError, CacheKeyId, NeverCanAuthor};
+use sp_consensus_aura::AuraApi;
+use sp_inherents::CreateInherentDataProviders;
+use sp_runtime::{
+    app_crypto::AppKey,
+    traits::{Block as BlockT, Header as HeaderT},
+};
+use std::{marker::PhantomData, sync::Arc};
+
+use nimbus_primitives::CompatibleDigestItem as NimbusDigestItem;
+use sc_consensus_aura::CompatibleDigestItem as AuraDigestItem;
+
+const LOG_TARGET: &str = "aura-nimbus-consensus";
+
+struct AuraOrNimbusVerifier<C, Block: BlockT, AuraCIDP, NimbusCIDP> {
+    aura_verifier:
+        sc_consensus_aura::AuraVerifier<C, <AuraId as AppKey>::Pair, NeverCanAuthor, AuraCIDP>,
+    nimbus_verifier: nimbus_consensus::Verifier<C, Block, NimbusCIDP>,
+}
+impl<C, Block, AuraCIDP, NimbusCIDP> AuraOrNimbusVerifier<C, Block, AuraCIDP, NimbusCIDP>
+where
+    Block: BlockT,
+{
+    pub fn new(
+        client: Arc<C>,
+        create_inherent_data_providers_aura: AuraCIDP,
+        create_inherent_data_providers_nimbus: NimbusCIDP,
+        telemetry: Option<TelemetryHandle>,
+    ) -> Self
+    where
+        C: ProvideRuntimeApi<Block> + Send + Sync + 'static,
+        <C as ProvideRuntimeApi<Block>>::Api: BlockBuilderApi<Block>,
+        AuraCIDP: CreateInherentDataProviders<Block, ()> + 'static,
+        NimbusCIDP: CreateInherentDataProviders<Block, ()> + 'static,
+    {
+        Self {
+            aura_verifier: sc_consensus_aura::build_verifier(BuildVerifierParams {
+                client: client.clone(),
+                create_inherent_data_providers: create_inherent_data_providers_aura,
+                // NOTE: We only support verification of historic aura blocks, not new block proposals using aura
+                can_author_with: NeverCanAuthor {},
+                check_for_equivocation: sc_consensus_aura::CheckForEquivocation::No,
```

Comment:
Why are we not checking for equivocation btw, and it's actually not clear who is checking who here

Reply:
Because Aura has no code actually *handling* equivocation beside a log print here: substrate/client/consensus/aura/src/import_queue.rs

Reply:
well, isn't that good enough, to at least spot suss behaviour, especially with the new log aggregation.

Reply:
oof yeah, nvm. I forgot that we're not using aura ... :D